libcontainer/configs/config_linux.go

```diff
@@ -22,14 +31,78 @@ type LinuxPersonality struct {
 	Domain int `json:"domain"`
 }
 
+// InsideContainer is set to true if we are running in the context of "runc
+// init". This must only be changed in startInitialization.
+//
+// TODO: Having this live in libcontainer/configs is quite ugly but due to the
+// import graph we can't really move it anywhere else (other than a dedicated
+// package).
+var InsideContainer bool
+
+func mappingFromNamespacePath(nsPath, mapName string) ([]IDMap, error) {
+	var mapData []byte
+	if InsideContainer {
+		// From within the container, we cannot join the target namespace (we
+		// don't have the necessary privileges, and we might be in the
+		// container's mountns after pivot_root(2) so the path might not exist
+		// either).
+		//
+		// If we happen to be inside the namespace in question, we can just
+		// read our own /proc/self/ file. However, calling HostUID and HostGID
+		// from the context of runc init should be avoided because there's no
+		// way for us to be sure that the namespace actually is the same as
+		// nsPath (and isn't different due to a programming bug). So emit a
+		// warning, but return our own mapping since that should still be
+		// correct in all cases.
+		logrus.Infof("cannot robustly retrieve mapping of namespace %s from within runc init, falling back to own mappings (this should work, but is probably a bug in runc or the libcontainer user -- please submit a bug report)", nsPath)
+		data, err := os.ReadFile("/proc/self/" + mapName)
+		if err != nil {
+			return nil, fmt.Errorf("retrieve runc init fallback mapping failed: %w", err)
+		}
+		mapData = data
+	} else {
+		// We have to join the namespace to get the actual mappings.
+		cmd := exec.Command("nsenter", "--no-fork", "--preserve-credentials", "--user="+nsPath, "--", "cat", "/proc/self/"+mapName)
+		output, err := cmd.Output()
+		if err != nil {
+			return nil, fmt.Errorf("getting /proc/self/%s from namespace %s failed: %s: %w", mapName, nsPath, bytes.TrimSpace(output), err)
+		}
+		mapData = output
+	}
```

Comment:
This is definitely ugly, and I suspect that @kolyshkin will want this to be cached (since we call `HostUID` and `HostGID` 5-6 times each container creation).

I guess we could add a cache to `Config` but it seems quite ugly to add runtime caching stuff to a structure we write to disk (obviously we can exclude it but it just looks ugly to me). It's also particularly ugly that this code is in `libcontainer/config` but if you want it to be a method on `Config` it has to live here (moving the code would introduce an import loop since we need the `Config` type in the code that does the conversion).

Ideally we could avoid all of this by removing the need for `Host*` wherever it is needed, unfortunately while there are very few users at least one of the users cannot be removed AFAICS. The only users of this code are:

```
libcontainer/container_linux.go:        rootuid, err := c.Config().HostRootUID()
libcontainer/container_linux.go:        rootgid, err := c.Config().HostRootGID()
libcontainer/specconv/spec_linux.go:    // config.HostRootUID() returns the correct result.
libcontainer/specconv/spec_linux.go:            ownerUid, err := config.HostUID(processUid)
libcontainer/specconv/spec_linux.go:    rootUID, err := config.HostRootUID()
libcontainer/specconv/spec_linux.go:    rootGID, err := config.HostRootGID()
utils_linux.go: rootuid, err := r.container.Config().HostRootUID()
utils_linux.go: rootgid, err := r.container.Config().HostRootGID()
```

Which, if you break it down are:

 1. Changing the owner of `exec.fifo`. Setting it to `0o666` would be a simple fix (since it's only used for signalling there are no security implications).
 2. Setting the `OwnerUID` of the cgroupfs when using cgroup namespaces.
 3. Changing the owners of all of the devices to root in the user namespace. This isn't necessary AFAICS because we cannot `mknod` inside user namespaces anyway, so mapping the owner doesn't serve any purpose.
 4. Setting the owner of io when using pipes for stdio. I wonder if setting the pipe to `0o666` would also work for this? Then again, while you cannot snoop on a process's file descriptors without privileges, I'm not sure how safe it is to change the mode of the underlying pipe...

AFACICS, usages 1 and 3 are easily removed but I'm not so sure about 4. And I don't see a way we could remove 2, because the semantics are in the runtime-spec...

Reply:
I think the `nsPath` is a string like this: `/proc/<pid>/ns/user`, why not get the pid value from `nsPath`, and read the content from `/proc/<pid>/(u|g)id_map'? Maybe there is some special conditions I missing?

Reply:
I think `config.HostRootUID` is only used by `runc`, so maybe we can move it to `spec_linux.go`, and cache it in `Container struct`?

Reply:
Ah, now that you mention it, we could probably cache this in the regular `config` fields when doing `specconv`. This would be much simpler and would also have the benefit of not needing to handle the "inside container" case (because we would do it when in the host and the data would be cached through exec since it's part of the configuration).

It is a little odd that we would have mappings and the path configuration, but it should be okay as long as the actual join logic (join the path if specified) stays the same. I will have to rework the changes to the validator (and we would have to validate that in specconv AFAICS).